rfc/src/rfcs/0009-function-contracts.md

```diff
@@ -0,0 +1,674 @@
+
+Verifying contrcts for recursive functions is performed inductively. Upon
+reentry of the function we use the replacement of the contract instead of the
+function body.
+
+The mechanism is generating an additional wrapper function around the check,
+with a global variable to track reentry. If we use the `pop` example from before
+we would get the following pattern after macro expansion.
+
+```rs
+#[checked_with = "recursion_wrapper"]
+#[replaced_with = "replace_pop"]
+fn pop(&mut self) { ... }
+
+fn check_pop(&mut self) { ... }
+
+fn replace_pop(&mut self) { ... }
+
+fn recursion_wrapper(&mut self) { 
+  static mut IS_ENTERED: bool;
+
+  if unsafe { IS_ENTERED } {
+    replace_pop(self)
+  } else {
+    unsafe { IS_ENTERED = true; }
+    check_pop(self)
+  }
+}
+```
+
+Note that this is insufficient to verify all types of recursive functions, as
+the contract specification language has no support for inductive lemmas
+inductive lemmas (for instance in [ACSL](https://frama-c.com/download/acsl.pdf)
+section 2.6.3 "inductive predicates"). Inductive lemmas are usually needed for
+recursive datastructures.
+
+### Changes to Other Components
+
+Contract enforcement and replacement (`kani::proof_for_contract(f)`,
+`kani::stub_verified(f)`) both dispatch to the **stubbing logic**, stubbing `f`
+with the generated check and replace function respectively. If `f` has no
+contract, an error is thrown.
+
+For **write sets and memory predicates** Kani relies on CBMC. Generated memory
+predicates (whether derived from types of from explicit clauses) are emitted
+from the compiler as GOTO contracts in the artifact. Then the driver invokes
+`goto-instrument` with the name of the GOTO-level function names to enforce or
+replace the memory contracts. The compiler communicates the names of the
+function via harness metadata.
+
+**Side effect** freedom is enforced with an MIR traversal over all code
+reachable from a contract expression. An error is thrown if known side-effecting
+actions are performed such as `ptr::write`, `malloc` or `free` and reported as a
+graceful compiler error. If the code calls out to C we rely on CBMC to catch the
+violation dynamically, resulting in a less readable error.
+
+<!-- 
+This is the technical portion of the RFC. Please provide high level details of the implementation you have in mind:
+
+- What are the main components that will be modified? (E.g.: changes to `kani-compiler`, `kani-driver`, metadata,
+  installation...)
+- How will they be modified? Any changes to how these components communicate?
+- Will this require any new dependency?
+- What corner cases do you anticipate? 
+-->
+
+## Rationale and alternatives
+
+<!-- For Developers -->
+<!-- `old` discussion here -->
+
+<!-- 
+- What are the pros and cons of this design?
+- What is the impact of not doing this?
+- What other designs have you considered? Why didn't you choose them? 
+-->
+
+
+- **Kani-side implementation vs CBMC** Instead of generating check and replace
```

Comment:
Each one of these items could be a section, and I think it'd help with readability.